confgenerator/otel/modular.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package otel provides data structures to represent and generate otel configuration.
+package otel
+
+import (
+	"fmt"
+
+	yaml "github.com/goccy/go-yaml"
+	"github.com/mitchellh/mapstructure"
+)
+
+// Pipeline represents a single OT receiver and zero or more processors that must be chained after that receiver.
+type Pipeline struct {
+	Receiver   Component
+	Processors []Component
+}
+
+// Component represents a single OT component (receiver, processor, exporter, etc.)
+type Component struct {
+	// Type is the string type needed to instantiate the OT component (e.g. "windowsperfcounters")
+	Type string
+	// Config is an object which can be serialized by mapstructure into the configuration for the component.
+	// This can either be a map[string]interface{} or a Config struct from OT.
+	Config interface{}
+}
+
+func (c Component) name(suffix string) string {
+	if suffix != "" {
+		return fmt.Sprintf("%s/%s", c.Type, suffix)
+	}
+	return c.Type
+}
+
+func configToYaml(config interface{}) ([]byte, error) {
+	outMap := make(map[string]interface{})
+	if err := mapstructure.Decode(config, &outMap); err != nil {
+		return nil, err
+	}
+	return yaml.Marshal(outMap)
+}
+
+type ModularReceiver interface {
+	Pipelines() []Pipeline
+}
+
+type ModularProcessor interface {
+	Processors() []Component
+}
+
+type ModularConfig struct {
+	Pipelines map[string]Pipeline
+	// GlobalProcessors and Exporter are added at the end of every pipeline.
+	// Only one instance of each will be created regardless of how many pipelines are defined.
+	GlobalProcessors []Component
+	Exporter         Component
+}
+
+func (c ModularConfig) Generate() (string, error) {
+	receivers := map[string]interface{}{}
+	processors := map[string]interface{}{}
+	exporters := map[string]interface{}{}
+	pipelines := map[string]interface{}{}
+
+	configMap := map[string]interface{}{
```

Comment:
A sample output would make this function more readable.  Something like:

    // metrics/mypipe:
    //   receivers: [hostmetrics/mypipe]
    //   processors: [filter/mypipe_1, metrics_filter/mypipe_2, resourcedetection/_global_0]
    //   exporters: [googlecloud]